### ISSUE TITLE
Rewrite _ss_environment to .env in 4.x upgrades

### ISSUE THREAD

**Issue by chillu** [silverstripe/silverstripe-upgrader#59]
# Acceptance Criteria

 * Runs as a command in upgrader
 * Renames all constants
 * Presents original and converted file back to user for review
 * Bails out of any conditional logic is detected, and points to helpful docs on how to implement those instead

# Notes

 * ~~Track under ORB-162~~ We've exhausted funding for that task.

# Pull Request
* [x] https://github.com/silverstripe/silverstripe-upgrader/pull/63
* [x] https://github.com/silverstripe/silverstripe-framework/pull/8011

**Comment by maxime-rainville** [comment]
# How to do this
* Create an EnvironementCommand under.
* Try to find the `_ss_environment.php`.
* Analyse the `_ss_environment.php` file to try to detect conditional logic.
  * If we detect conditional logic, we display warning and quite.
* Require the `_ss_environment.php`
  * use `get_defined_constants` to get all the constants.
  * read $_FILE_TO_URL_MAPPING and try to find an entry for the current path.
* Map old constant values to new values. Custom constants will be copied as-is.
* Write up the values to to a .env file.
* Give the option to delete `_ss_environment.php` file. (This file can be used by many sites, so we can't just blast it blindly)

# Understanding PHP
Upgrader already use [PHP](https://github.com/nikic/PHP-Parser). So it makes sense to keep using it for trying to detect conditional code.

[The doc for the parser](https://github.com/nikic/PHP-Parser/blob/master/doc/2_Usage_of_basic_components.markdown#node-tree-structure) says it can understand comments. I'll try to see if I can copy comments over to the `.env` file, but no promises.

# Security warning
This goes without saying, but we're running arbitrary code here. So if there's malicious code in the `_ss_environment.php` bad things will happen.

# Const equivalence between between SS4 and SS3
Most of SS3 constant have direct SS4 equivalent. Let me know if they are any SS3 constant not document in https://docs.silverstripe.org/en/3/getting_started/environment_management/

| SS4                       | SS3                             |
|---------------------------|---------------------------------|
| deprecated                | TEMP_FOLDER                     |
| SS_DATABASE_CLASS         | SS_DATABASE_CLASS               |
| SS_DATABASE_SERVER        | SS_DATABASE_SERVER              |
| SS_DATABASE_USERNAME      | SS_DATABASE_USERNAME            |
| SS_DATABASE_PASSWORD      | SS_DATABASE_PASSWORD            |
| SS_DATABASE_PORT          | SS_DATABASE_PORT                |
| SS_DATABASE_SUFFIX        | SS_DATABASE_SUFFIX              |
| SS_DATABASE_PREFIX        | SS_DATABASE_PREFIX              |
| SS_DATABASE_TIMEZONE      | SS_DATABASE_TIMEZONE            |
| SS_DATABASE_NAME          | SS_DATABASE_NAME                |
| SS_DATABASE_CHOOSE_NAME   | SS_DATABASE_CHOOSE_NAME         |
| SS_DEPRECATION_ENABLED    | SS_DEPRECATION_ENABLED          |
| SS_ENVIRONMENT_TYPE       | SS_ENVIRONMENT_TYPE             |
| SS_DEFAULT_ADMIN_USERNAME | SS_DEFAULT_ADMIN_USERNAME       |
| SS_DEFAULT_ADMIN_PASSWORD | SS_DEFAULT_ADMIN_PASSWORD       |
| SS_USE_BASIC_AUTH         | SS_USE_BASIC_AUTH               |
| SS_SEND_ALL_EMAILS_TO     | SS_SEND_ALL_EMAILS_TO           |
| SS_SEND_ALL_EMAILS_FROM   | SS_SEND_ALL_EMAILS_FROM         |
| SS_ERROR_LOG              | SS_ERROR_LOG                    |
| SS_PROTECTED_ASSETS_PATH  | No equivalent                   |
| SS_DATABASE_MEMORY        | No equivalent                   |
| SS_TRUSTED_PROXY_IPS      | No equivalent                   |
| SS_ALLOWED_HOSTS          | No equivalent                   |
| SS_MANIFESTCACHE          | No equivalent                   |
| SS_IGNORE_DOT_ENV         | No equivalent                   |
| SS_BASE_URL               | read from $_FILE_TO_URL_MAPPING |
| SS_CONFIGSTATICMANIFEST   | SS_CONFIGSTATICMANIFEST         |
| SS_DATABASE_SSL_KEY       | SS_DATABASE_SSL_KEY             |
| SS_DATABASE_SSL_CERT      | SS_DATABASE_SSL_CERT            |
| SS_DATABASE_SSL_CA        | SS_DATABASE_SSL_CA              |
| SS_DATABASE_SSL_CIPHER    | SS_DATABASE_SSL_CIPHER          |

# Other questions and remarks
Should I try to remove `require_once("conf/ConfigureFromEnv.php");` from `_config.php`?

Also, normally an environment file shouldn't be track through git. Do we want to create a `.env.sample` file?

**Comment by tractorcow** [comment]
As far as the env mapping is concerned, I don't think we need to worry too much about mapping directly. The only custom logic I would bake in is the file-url mapping.

If you are including the file directly, we can rewrite vars even if there is conditional logic, since those conditions would be evaluated. It only needs to be a warning in that case isn't it?

**Comment by sminnee** [comment]
Can't we just rewrite each define to an environment variable of the same name, with a few exceptions?

**Comment by maxime-rainville** [comment]
@tractorcow Just giving a warning could work. It really goes down to how much we want the developer to be aware of what the upgrader is doing. It doesn't really affect the amount of work required either way and it would be easy to change our mind later. 

@sminnee The issue is that there's some `define` that might be called or not be called based on some other condition. So I can't just search for every instance of `define` and blindly copy the values over to a .env file.

That's an example from the official doc.
```php
if(isset($_SERVER['REMOTE_ADDR']) && ($_SERVER['REMOTE_ADDR'] == '127.0.0.1' || ($_SERVER['REMOTE_ADDR'] == '<MyIP>' 
&& strpos($_SERVER['HTTP_USER_AGENT'], '+') !== false))) 
    define('SS_ENVIRONMENT_TYPE', 'dev');
else 
    define('SS_ENVIRONMENT_TYPE', 'live');
```

**Comment by sminnee** [comment]
Official docs are out of date, then – SS_ENVIRONMENT_TYPE as a define is wrong

Also that code's a bad idea :P

**Comment by maxime-rainville** [comment]
@sminnee I know it's a bad idea. But I get a feeling there's probably a lot of `_ss_environment.php` file out there full of bad ideas.

**Comment by maxime-rainville** [comment]
Speaking of bad ideas. 

I'm trying to think of way of reading the constant defined in the `_ss_environment.php` file from the upgrader. The two ways I can think off are: 
* to include/eval the code directly in my own code. e.g.: 
```php
include($path_to_ss_environment);
# or
eval($content_of_ss_environement);
```
* to exec the content of `_ss_environment.php` and check the output. e.g.: 
```php
$phpcode = $content_of_ss_environement . $some_code_serialize_constants; 
exec("php -r " . $phpcode, $output);
```

Solution 1 will pollute the upgrader with a bunch of constants. Solution 2 feels dodgy.

Can someone think of a better way?

**Comment by martinduparc** [comment]
This command doesn't seem to be part of the V1 release (`Command "environment" is not defined.`).

**Comment by maxime-rainville** [comment]
@martinduparc It hasn't been release yet. It will be available with version 1.1.

If you want to use right away, you can install the `dev/master` branch of the upgrader with ` composer global require silverstripe/upgrader dev-master`. Obviously, keep in mind that you will be using software under active development. Feel free to log an issue if you come across any bugs.
